### PR TITLE
Change variable validation error message type from string to expression

### LINKED ISSUE #24160: More powerful error message generation with custom variable validation
### Current Terraform Version

```
Terraform v0.12.20
```

### Use-cases

When validating lists, sets or maps using custom variable validation, it would be helpful to be able to reference which item in the collection is invalid from within the error message. Imagine adding a new validation to a list with 20 items and trying to figure out which one is invalid.

### Attempted Solutions

I thought that I could essentially re-do the validation inside a string interpolation block, such as this:

```hcl
validation {
  condition     = length([for topic in var.topics : topic.name if length(topic.acls) == 0]) == 0
  error_message = "Topics must have at least one ACL attached. Invalid topic(s): ${join(", ", [for topic in var.topics : topic.name if length(topic.acls) == 0])}"
}
```
But this yielded the error message:

```
Error: Function calls not allowed

  on ../../modules/confluent-cloud/kafka/variables.tf line 92, in variable "topics":
  92:     error_message = "Topics must have at least one ACL attached. Invalid topic(s): ${join(", ", [for topic in var.topics : topic.name if length(topic.acls) == 0])}"

Functions may not be called here.
```

### Proposal

Allow function calls within the error message so that we can generate more helpful errors.

### LINKED ISSUE #24407: Variable Validation: allow functions in error_message
Here's a validation test I wrote for a MySQL username:

```hcl
validation {
    condition     = length(var.username) > 16
    error_message = "The username cannot be more than 16 characters (first 16 are ${trim(var.username, 0, 16)})."
  }
```

I wanted to give a hint to the operator about the longest string allowed, to make their life easier. But you can't use functions in `error_message`. Please add support for the above use-case.

### PR DESCRIPTION
**Pull request by joeaawad**
Submitted on behalf of a third-party: @cgetzen

All credit goes to @cgetzen for https://github.com/hashicorp/terraform/pull/25028 which originally made this change almost 10 months ago. I just updated it to resolve the merge conflicts since the original PR appears to be abandoned and I think this is a vital PR that would make a huge quality of life improvement for users and save me a lot of time answering questions from less experienced Terraform users.

Resolves #24407.
Resolves #24160.

This PR changes configs.VariableValidation.ErrorMessage from a string to an expression.

One consequence of this is that some of the validations on the error message (nilstring and lookslikesentence) are moved from the configs package to the terraform package and checked during Eval.

#### Template
This file is used as a template for the results below

```terraform
variable "x" {
  default = "four"
  type    = string
  validation {
    condition     = length(var.x) > 5
    error_message = "not a sentence"
  }
}
```

#### error_message 
```
output
```
#### "Input '${var.x}' must be more than 5 characters."
```
Error: Invalid value for variable

  on main.tf line 6, in module "test":
   6:   x = each.key

Input 'one' must be more than 5 characters.

This was checked by the validation rule at module/variables.tf:4,3-13.


Error: Invalid value for variable

  on main.tf line 6, in module "test":
   6:   x = each.key

Input 'two' must be more than 5 characters.

This was checked by the validation rule at module/variables.tf:4,3-13.
```
#### [4, 3]   
```
Error: Invalid validation error message

  on module/variables.tf line 6, in variable "x":
   6:     error_message = [4, 3]

Invalid validation error message result value: string required.
``` 

#### "Your string ${var.x} isn't long enough." 
```
Error: Invalid value for variable

  on main.tf line 5, in module "test":
   5:   x = "four"

Your string four isn't long enough.

This was checked by the validation rule at module/variables.tf:4,3-13.
```

#### "not a sentence"  
```
Error: Invalid value for variable

  on main.tf line 5, in module "test":
   5:   x = "four"

not a sentence

This was checked by the validation rule at module/variables.tf:4,3-13.


Error: Invalid validation error message

  on module/variables.tf line 1:
   1: variable "x" {

Validation error message must be at least one full English sentence starting with an uppercase letter and ending with a period
or question mark.
```

### CODE REVIEW

File: terraform/eval_variable.go
Comment (joeaawad) [inline terraform/eval_variable.go:29]: `n.Expr` appears to have been changed to `expr` a while back so this is just fixing the docstring since I noticed it when fixing the merge conflicts from the original PR

**Comment by cgetzen** [comment]
Hey @joeaawad, glad to see others have the same language requests. My PR was waiting on a review, but if you're able to get the required +1s you need on this I'll go ahead and close mine. Thanks!

**Comment by joeaawad** [comment]
Sounds good, thanks @cgetzen. Didn't mean to step on any toes, just saw the merge conflicts and lack of activity so was hoping to try to push this through in time for 0.15.0. Maintainers - is 0.15.0 a possibility or is that already locked? Happy to help in any way that I can to push this through in time.

**Comment by joeaawad** [comment]
@pkolyvas Does https://discuss.hashicorp.com/t/terraform-0-15-0-beta2-released/21368 mean that it is too late to get this change in for `0.15.0`? How can I help push this along to get it out ASAP?

**Review by Hem-Bhatt — APPROVED** [review]

**Review by Hem-Bhatt — APPROVED** [review]

**Comment by joeaawad** [comment]
@apparentlymart @alisdair this was pretty skewered by https://github.com/hashicorp/terraform/pull/30401 refactor that lost the EvalContext so I'm pretty out of my depth here on how to get a string out of [`cr.ErrorMessage`](https://github.com/hashicorp/terraform/blob/main/internal/configs/checks.go#L119-L126) after the [type of ErrorMessage](https://github.com/hashicorp/terraform/blob/main/internal/configs/checks.go#L32) is changed to `hcl.Expression` for the [comparisons](https://github.com/hashicorp/terraform/pull/28044/files#diff-c5efa377b938a12783a77cc1a94415484dd592bf37d7cc7182563ecfb0024b95R168-R175). Maybe it would make sense to add a new `ErrorMessageExpr` like the [ConditionalExpr](https://github.com/hashicorp/hcl2/blob/master/hcl/hclsyntax/expression.go#L457-L612) to [hclsyntax/expression](https://github.com/hashicorp/hcl2/blob/master/hcl/hclsyntax/expression.go) and have that produce the string? 

Please advise on if there is a path forwards for me to do this or if it would be best handled by the HashiCorp team, but I do feel that there is a ton of value in pushing this forward, especially now that you are looking at pre and post conditions which could also benefit from being able to explain exactly what variable caused the error.

**Comment by alisdair** [comment]
Hi @joeaawad! Thanks for working on this, and I'm sorry I hadn't noticed it until now.

I agree with your assessment of the value in evaluating error messages as expressions, especially if we move forward with pre/postconditions or something similar. I had actually been working on a separate implementation of the same concept myself, as I hadn't seen this branch or PR. I'll be opening a pull request for that shortly, and if it's approved then I hope we'll have error message expressions in the 1.2 release.

I'm going to close this for now in favour of the other branch, rather than as you to do further work on this one. Thanks again for pushing this forward!

**Comment by joeaawad** [comment]
@alisdair awesome, glad to hear it, thanks a lot

**Pull request closed without merging**
